### PR TITLE
Allow JDBC connector to invalidate statistics cache

### PR DESCRIPTION
**Pull request by kokosing**
Allow JDBC connector to invalidate statistics cache

### CODE REVIEW

**Comment by findepi** [comment]
Why don't you expose cache eviction (for everything) as a JMX operation instead?

**Comment by kokosing** [comment]
> Why don't you expose cache eviction (for everything) as a JMX operation instead?

It is not dedicated for user. It is dedicated for a connector.

Just sit back and relax. Close your eyes and image that out there in the universe there could be a JDBC connector that is capable to capture relation statistics during the table scan. Such connector would be delighted being able to invalidate cached statistics, so user could be blessed by fresh and warm statistics.

**Review by ssheikin — APPROVED** [review]

**Comment by kokosing** [comment]
CI hit: https://github.com/trinodb/trino/issues/7534

File: plugin/trino-base-jdbc/src/main/java/io/trino/plugin/jdbc/remote/RemoteEventManager.java [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+import com.google.common.collect.ImmutableSet;
+import io.trino.plugin.jdbc.JdbcTableHandle;
+
+import javax.inject.Inject;
+
+import java.util.Set;
+
+import static java.util.Objects.requireNonNull;
+
+public class RemoteEventManager
```
Comment (findepi) [inline plugin/trino-base-jdbc/src/main/java/io/trino/plugin/jdbc/remote/RemoteEventManager.java:25]: `com.google.common.eventbus.EventBus` ?

**Comment by kokosing** [comment]
CI hit https://github.com/trinodb/trino/issues/7619 and https://github.com/trinodb/trino/issues/7534

File: plugin/trino-base-jdbc/src/main/java/io/trino/plugin/jdbc/JdbcModule.java [hunk 2]
```diff
@@ -80,6 +80,20 @@ public ConnectionFactory createConnectionFactory(@StatsCollecting ConnectionFact
         return new LazyConnectionFactory(connectionFactory);
     }
 
+    @Provides
+    @Singleton
+    public JdbcClient createJdbcClient(CachingJdbcClient cachingJdbcClient)
```
Comment (findepi) [inline plugin/trino-base-jdbc/src/main/java/io/trino/plugin/jdbc/JdbcModule.java:85]: what is this for?

File: plugin/trino-base-jdbc/src/main/java/io/trino/plugin/jdbc/JdbcModule.java [hunk 3]
```diff
@@ -80,6 +80,20 @@ public ConnectionFactory createConnectionFactory(@StatsCollecting ConnectionFact
...
+    {
+        return cachingJdbcClient;
+    }
+
+    @Provides
+    @Singleton
+    public RemoteEventSink createRemoteEventSink(CachingJdbcClient cachingJdbcClient)
```
Comment (findepi) [inline plugin/trino-base-jdbc/src/main/java/io/trino/plugin/jdbc/JdbcModule.java:92]: why not 
```
binder.bind(RemoteEventSink.class).to(CachingJdbcClient.class)
```
?

File: plugin/trino-base-jdbc/src/main/java/io/trino/plugin/jdbc/RemoteEventSink.java [hunk 4]
```diff
@@ -0,0 +1,22 @@
...
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.trino.plugin.jdbc;
+
+/**
+ * A sink of remote data source events that out of the control of Trino.
+ */
+public interface RemoteEventSink
```
Comment (findepi) [inline plugin/trino-base-jdbc/src/main/java/io/trino/plugin/jdbc/RemoteEventSink.java:19]: i don't think this interfce gives us much without an event bus

File: plugin/trino-base-jdbc/src/main/java/io/trino/plugin/jdbc/CachingJdbcClient.java [hunk 5]
```diff
@@ -497,6 +497,12 @@ CacheStats getStatisticsCacheStats()
         cache.invalidateAll(cacheKeys);
     }
 
+    @Override
+    public void fireDataChangedEvent(JdbcTableHandle handle)
```
Comment (findepi) [inline plugin/trino-base-jdbc/src/main/java/io/trino/plugin/jdbc/CachingJdbcClient.java:501]: or `onDataChange` 

("fire" seems more appropriate to event dispatcher to me)

**Review by findepi — APPROVED** [review]
optional nits

File: plugin/trino-base-jdbc/src/main/java/io/trino/plugin/jdbc/CachingJdbcClient.java (hunk 5 above)
Comment (kokosing) [reply plugin/trino-base-jdbc/src/main/java/io/trino/plugin/jdbc/CachingJdbcClient.java:501]: "fire" and "on" are good prefixes for event bus, where one side triggers and other side responds. Here things are bundled together so neither of these two fits.

File: plugin/trino-base-jdbc/src/main/java/io/trino/plugin/jdbc/CachingJdbcClient.java [hunk 6]
```diff
@@ -497,6 +497,11 @@ CacheStats getStatisticsCacheStats()
         cache.invalidateAll(cacheKeys);
     }
 
+    public void onDataChanged(JdbcTableHandle handle)
```
Comment (findepi) [inline plugin/trino-base-jdbc/src/main/java/io/trino/plugin/jdbc/CachingJdbcClient.java:500]: `JdbcTableHandle` is not a good representation of the table. For example, we don't want to distinguish between "a plan table" and "table with selected columns", or "a table with a constraint" here.

Comment (findepi) [reply plugin/trino-base-jdbc/src/main/java/io/trino/plugin/jdbc/CachingJdbcClient.java:500]: https://github.com/trinodb/trino/pull/8237

Comment (kokosing) [reply plugin/trino-base-jdbc/src/main/java/io/trino/plugin/jdbc/CachingJdbcClient.java:500]: That was the point. To allow invalidate only partial information about the table.

**Pull request merged**